### PR TITLE
feat: adds interaction and dialog node types

### PR DESCRIPTION
**Pull request by balloonpopper**

### CODE REVIEW

**Comment by BHSDuncan** [comment]
I agree with @StraToN . Of the two proposals he put forth, I'd tend to favour the first one as it promotes composition and avoids adding another property.

**Comment by balloonpopper** [comment]
@StraToN / @BHSDuncan - Based on your request, I've created a new node types for interaction locations, inheriting from ESCLocation. It didn't make sense to leave the existing logic in that looks for a node specifically called "dialog_location" if the plan was not to use a named node for interaction positions, so I've created a dialog_location node type as well. The Mark and Worker characters have been updated to use the new dialog location node type. 

For ESCItems, if an "ESCInteractionLocation" node isn't found, it falls back to an "ESCLocation" node, which falls back to the middle of the collision shape.  If it doesn't have a collision shape, the click on the item is never registered, so the get_interact_position() function is never called.

File: addons/escoria-core/game/core-scripts/esc_item.gd [hunk 1]
```diff
@@ -149,6 +149,10 @@ export(NodePath) var animation_player_node: NodePath = "" \
 # as a camera target
 export(NodePath) var camera_node
 
+# Custom data dictionary to ease customization and custom command creation.
+# Avoid name collision using proper key names.
+export(Dictionary) var custom_data = {}
```
Comment (StraToN) [inline addons/escoria-core/game/core-scripts/esc_item.gd:154]: I think you included code from PR#705, which was merged a bit earlier. You may end up with conflicts here. `Git rebase` is probably needed here ;)

File: addons/escoria-dialog-simple/types/floating.gd [hunk 2]
```diff
@@ -142,10 +161,12 @@ func say(character: String, line: String) :
 		.format([text_color_html]) + tr(line) + "[/color][center]"
 
 	if _current_character.is_inside_tree() and \
-			_current_character.has_node("dialog_position"):
-		rect_position = _current_character.get_node(
-			"dialog_position"
-		).get_global_transform_with_canvas().origin
+			dialog_location_node != null:
+#		rect_position = _current_character.get_node(
```
Comment (StraToN) [inline addons/escoria-dialog-simple/types/floating.gd:165]: Is this commented block still needed?

**Review by StraToN — CHANGES_REQUESTED** [review]

File: addons/escoria-core/game/core-scripts/esc_item.gd [hunk 3]
```diff
@@ -516,31 +520,53 @@ func get_animation_player() -> Node:
...
+				interact_count += 1
+				interact_position = c.global_position
+			else:
+				# This will catch all other Position2D related nodes
+				# including dialog locations and native Position2D nodes.
+				pos_2d_count += 1
+				pos_2d_position = c.global_position				
+	
+	if interact_position == null and \
+	esclocation_position == null and collision != null:
```
Comment (BHSDuncan) [inline addons/escoria-core/game/core-scripts/esc_item.gd:548]:
```suggestion
		esclocation_position == null and collision != null:
```

File: addons/escoria-core/game/core-scripts/esc_item.gd [hunk 4]
```diff
@@ -516,31 +520,53 @@ func get_animation_player() -> Node:
...
+				"Multiple ESCInteractionLocations found to walk to for " +
+				"object %s. Last one will be used." % global_id)
+		return interact_position
+	else:
+		if esclocation_count > 1:
+			escoria.logger.warn(
+				self,
+				"Multiple ESClocations found to walk to for object " +
+				"%s. Last one will be used." % global_id)
+		return esclocation_position
```
Comment (BHSDuncan) [inline addons/escoria-core/game/core-scripts/esc_item.gd:568]:
```suggestion

	if esclocation_count > 1:
		escoria.logger.warn(
			self,
			"Multiple ESClocations found to walk to for object " +
			"%s. Last one will be used." % global_id)
	return esclocation_position
```

File: addons/escoria-dialog-simple/types/floating.gd [hunk 5]
```diff
@@ -105,13 +106,15 @@ func _ready():
 
 	_current_line = ""
 
+	
 
 func _process(delta):
 	if _current_character.is_inside_tree() and \
-			_current_character.has_node("dialog_position"):
+			dialog_location_node != null:
```
Comment (BHSDuncan) [inline addons/escoria-dialog-simple/types/floating.gd:113]:
```suggestion
			is_instance_valid(dialog_location_node):
```

File: addons/escoria-dialog-simple/types/floating.gd [hunk 6]
```diff
@@ -142,10 +161,12 @@ func say(character: String, line: String) :
 		.format([text_color_html]) + tr(line) + "[/color][center]"
 
 	if _current_character.is_inside_tree() and \
-			_current_character.has_node("dialog_position"):
-		rect_position = _current_character.get_node(
-			"dialog_position"
-		).get_global_transform_with_canvas().origin
+			dialog_location_node != null:
```
Comment (BHSDuncan) [inline addons/escoria-dialog-simple/types/floating.gd:164]:
```suggestion
			is_instance_valid(dialog_location_node):
```

**Review by BHSDuncan — CHANGES_REQUESTED** [review]

File: addons/escoria-core/game/core-scripts/esc_item.gd (hunk 1 above)
Comment (balloonpopper) [reply addons/escoria-core/game/core-scripts/esc_item.gd:154]: This codeblock is in the current develop branch - https://github.com/godot-escoria/escoria-demo-game/blob/develop/addons/escoria-core/game/core-scripts/esc_item.gd#L154 ?

File: addons/escoria-dialog-simple/types/floating.gd (hunk 2 above)
Comment (balloonpopper) [reply addons/escoria-dialog-simple/types/floating.gd:164]: Oops, thanks! Fixed. Also found another comment I'd left in and fixed that too.

File: addons/escoria-core/game/core-scripts/esc_item.gd (hunk 3 above)
Comment (balloonpopper) [reply addons/escoria-core/game/core-scripts/esc_item.gd]: Only the "collision" works with "is_instance_valid", the other two are coordinates. Updated to "! is_instanace_valid(collision)"

File: addons/escoria-core/game/core-scripts/esc_interact_location.gd
[the remark this answers is not in the document]
Comment (balloonpopper) [reply addons/escoria-core/game/core-scripts/esc_interact_location.gd:1]: Done

**Review by BHSDuncan — APPROVED** [review]

**Comment by BHSDuncan** [comment]
Good work!

**Review by StraToN — APPROVED** [review]

**Comment by StraToN** [comment]
Great work, thank you! 🥳

**Pull request merged**